tb_plugin/torch_tb_profiler/profiler/overall_parser.py

```diff
@@ -274,6 +332,139 @@ def update_steps_consider_device_side(self, runtime_node_list, device_node_list)
+                for i_bucket in range(buckets):
+                    gpu_utilization_timeline[-1][i_bucket] /= bucket_size
+
+                for i_bucket in range(buckets):
+                    start_time = self.steps[0][0] + i_bucket * bucket_size
+                    util_json = build_trace_counter(device_id, start_time, gpu_utilization_timeline[-1][i_bucket])
+                    counter_json.append(util_json)
+                start_time = self.steps[0][0] + buckets * bucket_size
+                util_json = build_trace_counter(device_id, start_time, 0)
+                counter_json.append(util_json)
+
+        self.kernel_ranges_per_device = None  # Release memory.
+
+        return counter_json
+
+
+    def calculate_approximated_sm_efficency(self):
+        def calculate_avg(approximated_sm_efficency_ranges, total_dur):
+            total_weighted_sm_efficiency = 0.0
+            for r in approximated_sm_efficency_ranges:
+                dur = r[0][1] - r[0][0]
+                total_weighted_sm_efficiency += r[1] * dur
+            avg_approximated_sm_efficency = total_weighted_sm_efficiency / total_dur
+            return avg_approximated_sm_efficency
+
+        def build_trace_counter(gpu_id, start_time, counter_value):
+            util_json = {}
+            util_json["ph"] = "C"
+            util_json["name"] = "GPU {} Est. SM Efficiency".format(gpu_id)
+            util_json["pid"] = 0
+            util_json["ts"] = start_time
+            util_json["args"] = {"Est. SM Efficiency": counter_value}
+            return util_json
+
+        counter_json = []
+        total_dur = self.steps[-1][1] - self.steps[0][0]
+        approximated_sm_efficency_ranges_per_device = []
+        for device_id, i in self.device_to_index.items():
+            blocks_per_sm_ranges = self.blocks_per_sm_per_device[i]
+            approximated_sm_efficency_ranges = merge_ranges_with_value(blocks_per_sm_ranges)
+            approximated_sm_efficency_ranges_per_device.append(approximated_sm_efficency_ranges)
+            avg_approximated_sm_efficency = calculate_avg(approximated_sm_efficency_ranges, total_dur)
+            self.avg_approximated_sm_efficency_per_device.append(avg_approximated_sm_efficency)
+
+            if avg_approximated_sm_efficency > 0.0:
+                for r in approximated_sm_efficency_ranges:
+                    efficiency_json_start = build_trace_counter(device_id, r[0][0], r[1])
+                    efficiency_json_finish = build_trace_counter(device_id, r[0][1], 0)
+                    counter_json.append(efficiency_json_start)
+                    counter_json.append(efficiency_json_finish)
+
+        self.blocks_per_sm_per_device = None  # Release memory.
+
+        return counter_json
+
+
+    # Weighted average. Weighted by kernel's time duration.
+    def calculate_occupancy(self):
+        for i in range(len(self.occupancy_per_device)):
```

Comment:
should the for occupancy in occupancy_per_device be enough?

Reply:
Remove the "i"? Good style. Done. Thanks.